sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegen.scala

```diff
@@ -116,6 +119,39 @@ trait CodegenSupport extends SparkPlan {
   }
 
   /**
+    * Returns source code to evaluate all the variables, and clear the code of them, to prevent
```

Comment:
I was imagining something like:

evaluateAttributes(Seq[Expression]) which evaluates all the attribute refernces in the tree that haven't been. This is kind of similar to what you have below.